File: apps/cpp/simulator/chemistry.cpp
#include "chemistry.hpp"

#include <algorithm>
#include <cmath>
#include <random>

namespace sim {

static thread_local std::mt19937& rng() {
    static thread_local std::mt19937 gen{456};
    return gen;
}

static double randUniform() {
    static thread_local std::uniform_real_distribution<double> dist(0.0, 1.0);
    return dist(rng());
}

// ============================================================
// ChemicalSystem
// ============================================================

ChemicalSystem::ChemicalSystem(AtomicSystem& as)
    : atomic(as) {}

int ChemicalSystem::nextMoleculeId() { return ++nextId_; }

std::vector<size_t> ChemicalSystem::findUnbonded(int z) const {
    std::vector<size_t> result;
    for (size_t i = 0; i < atomic.atoms.size(); ++i) {
        if (atomic.atoms[i].atomicNumber == z && atomic.atoms[i].bonds.empty())
            result.push_back(i);
    }
    return result;
}

std::vector<size_t> ChemicalSystem::findWithBondsBelow(int z, size_t maxBonds) const {
    std::vector<size_t> result;
    for (size_t i = 0; i < atomic.atoms.size(); ++i) {
        if (atomic.atoms[i].atomicNumber == z && atomic.atoms[i].bonds.size() < maxBonds)
            result.push_back(i);
    }
    return result;
}

std::vector<Molecule> ChemicalSystem::formWater() {
    std::vector<Molecule> waters;

    auto hydrogens = findUnbonded(1);
    auto oxygens   = findWithBondsBelow(8, 2);

    while (hydrogens.size() >= 2 && !oxygens.empty()) {
        size_t hi1 = hydrogens.back(); hydrogens.pop_back();
        size_t hi2 = hydrogens.back(); hydrogens.pop_back();
        size_t oi  = oxygens.back();   oxygens.pop_back();

        auto& h1 = atomic.atoms[hi1];
        auto& h2 = atomic.atoms[hi2];
        auto& o  = atomic.atoms[oi];

        Molecule water;
        water.moleculeId = nextMoleculeId();
        water.name = "water";
        water.formula = "H2O";
        water.position = o.position;
        water.atomIds = {h1.atomId, h2.atomId, o.atomId};

        h1.bonds.push_back(o.atomId);
        h2.bonds.push_back(o.atomId);
        o.bonds.push_back(h1.atomId);
        o.bonds.push_back(h2.atomId);

        waters.push_back(water);
        molecules.push_back(water);
        ++waterCount;
    }

    return waters;
}

std::vector<Molecule> ChemicalSystem::formMethane() {
    std::vector<Molecule> methanes;

    auto carbons   = findUnbonded(6);
    auto hydrogens = findUnbonded(1);

    while (!carbons.empty() && hydrogens.size() >= 4) {
        size_t ci = carbons.back(); carbons.pop_back();

        std::vector<size_t> hIndices;
        for (int k = 0; k < 4; ++k) {
            hIndices.push_back(hydrogens.back());
            hydrogens.pop_back();
        }

        auto& c = atomic.atoms[ci];
        Molecule methane;
        methane.moleculeId = nextMoleculeId();
        methane.name = "methane";
        methane.formula = "CH4";
        methane.position = c.position;
        methane.isOrganic = true;
        methane.atomIds.push_back(c.atomId);

        for (auto hi : hIndices) {
            auto& h = atomic.atoms[hi];
            c.bonds.push_back(h.atomId);
            h.bonds.push_back(c.atomId);
            methane.atomIds.push_back(h.atomId);
        }

        methanes.push_back(methane);
        molecules.push_back(methane);
    }

    return methanes;
}

std::vector<Molecule> ChemicalSystem::formAmmonia() {
    std::vector<Molecule> ammonias;

    auto nitrogens = findUnbonded(7);
    auto hydrogens = findUnbonded(1);

    while (!nitrogens.empty() && hydrogens.size() >= 3) {
        size_t ni = nitrogens.back(); nitrogens.pop_back();

        std::vector<size_t> hIndices;
        for (int k = 0; k < 3; ++k) {
            hIndices.push_back(hydrogens.back());
            hydrogens.pop_back();
        }

        auto& n = atomic.atoms[ni];
        Molecule ammonia;
        ammonia.moleculeId = nextMoleculeId();
        ammonia.name = "ammonia";
        ammonia.formula = "NH3";
        ammonia.position = n.position;
        ammonia.atomIds.push_back(n.atomId);

        for (auto hi : hIndices) {
            auto& h = atomic.atoms[hi];
            n.bonds.push_back(h.atomId);
            h.bonds.push_back(n.atomId);
            ammonia.atomIds.push_back(h.atomId);
        }

        ammonias.push_back(ammonia);
        molecules.push_back(ammonia);
    }

    return ammonias;
}

bool ChemicalSystem::formAminoAcid(const std::string& aaType) {
    // Glycine-like: 2C + 5H + 2O + 1N minimum
    auto carbons   = findUnbonded(6);
    auto hydrogens = findUnbonded(1);
    auto oxygens   = findWithBondsBelow(8, 2);
    auto nitrogens = findUnbonded(7);

    if (carbons.size() < 2 || hydrogens.size() < 5 ||
        oxygens.size() < 2 || nitrogens.empty())
        return false;

    Molecule aa;
    aa.moleculeId = nextMoleculeId();
    aa.name = aaType;
    aa.isOrganic = true;
    aa.functionalGroups = {"amino", "carboxyl"};

    // Use atoms: 2C, 5H, 2O, 1N
    std::vector<size_t> used;
    for (int k = 0; k < 2; ++k) { used.push_back(carbons.back()); carbons.pop_back(); }
    for (int k = 0; k < 5; ++k) { used.push_back(hydrogens.back()); hydrogens.pop_back(); }
    for (int k = 0; k < 2; ++k) { used.push_back(oxygens.back()); oxygens.pop_back(); }
    used.push_back(nitrogens.back()); nitrogens.pop_back();

    aa.position = atomic.atoms[used[0]].position;
    int firstId = atomic.atoms[used[0]].atomId;
    for (auto idx : used) {
        auto& a = atomic.atoms[idx];
        aa.atomIds.push_back(a.atomId);
        // Bond all to the first atom (simplified)
        if (a.atomId != firstId) {
            atomic.atoms[used[0]].bonds.push_back(a.atomId);
            a.bonds.push_back(firstId);
        }
    }

    molecules.push_back(aa);
    ++aminoAcidCount;
    return true;
}

bool ChemicalSystem::formNucleotide(const std::string& base) {
    // Simplified: 5C + 8H + 4O + 2N
    auto carbons   = findUnbonded(6);
    auto hydrogens = findUnbonded(1);
    auto oxygens   = findWithBondsBelow(8, 2);
    auto nitrogens = findUnbonded(7);

    if (carbons.size() < 5 || hydrogens.size() < 8 ||
        oxygens.size() < 4 || nitrogens.size() < 2)
        return false;

    Molecule nuc;
    nuc.moleculeId = nextMoleculeId();
    nuc.name = "nucleotide-" + base;
    nuc.isOrganic = true;
    nuc.functionalGroups = {"sugar", "phosphate", "base"};

    std::vector<size_t> used;
    for (int k = 0; k < 5; ++k) { used.push_back(carbons.back()); carbons.pop_back(); }
    for (int k = 0; k < 8; ++k) { used.push_back(hydrogens.back()); hydrogens.pop_back(); }
    for (int k = 0; k < 4; ++k) { used.push_back(oxygens.back()); oxygens.pop_back(); }
    for (int k = 0; k < 2; ++k) { used.push_back(nitrogens.back()); nitrogens.pop_back(); }

    nuc.position = atomic.atoms[used[0]].position;
    int firstId = atomic.atoms[used[0]].atomId;
    for (auto idx : used) {
        auto& a = atomic.atoms[idx];
        nuc.atomIds.push_back(a.atomId);
        if (a.atomId != firstId) {
            atomic.atoms[used[0]].bonds.push_back(a.atomId);
            a.bonds.push_back(firstId);
        }
    }

    molecules.push_back(nuc);
    ++nucleotideCount;
    return true;
}

int ChemicalSystem::catalyzedReaction(double temperature, bool catalystPresent) {
    int formed = 0;
    double eaFactor = catalystPresent ? 0.3 : 1.0;
    double thermal  = K_B * temperature;

    // Amino acids
    if (thermal > 0.0 && atomic.atoms.size() > 10) {
        double aaProb = std::exp(-5.0 * eaFactor / (thermal + 1e-20));
        if (randUniform() < aaProb) {
            int idx = static_cast<int>(randUniform() * AMINO_ACIDS.size());
            if (idx >= static_cast<int>(AMINO_ACIDS.size())) idx = 0;
            if (formAminoAcid(std::string(AMINO_ACIDS[idx]))) {
                ++formed;
                ++reactionsOccurred;
            }
        }
    }

    // Nucleotides
    if (thermal > 0.0 && atomic.atoms.size() > 19) {
        double nucProb = std::exp(-8.0 * eaFactor / (thermal + 1e-20));
        if (randUniform() < nucProb) {
            const char* bases[] = {"A", "T", "G", "C"};
            int idx = static_cast<int>(randUniform() * 4);
            if (idx >= 4) idx = 0;
            if (formNucleotide(bases[idx])) {
                ++formed;
                ++reactionsOccurred;
            }
        }
    }

    return formed;
}

} // namespace sim

File: apps/php/simulator/Chemistry.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/Constants.php';
require_once __DIR__ . '/Atomic.php';

/**
 * Chemistry simulation - molecular assembly and reactions.
 *
 * Models formation of molecules from atoms: water, amino acids,
 * nucleotides, and other biomolecules essential for life.
 * Chemical reactions are driven by energy, catalysis, and concentration.
 */

class Molecule
{
    private static int $idCounter = 0;

    public readonly int $moleculeId;
    public string $formula;
    public bool $isOrganic;

    /**
     * @param Atom[]   $atoms
     * @param string   $name
     * @param float[]  $position
     * @param float    $energy
     * @param string[] $functionalGroups
     */
    public function __construct(
        public array $atoms = [],
        public string $name = '',
        ?string $formula = null,
        ?array $position = null,
        public float $energy = 0.0,
        bool $isOrganic = false,
        public array $functionalGroups = [],
    ) {
        self::$idCounter++;
        $this->moleculeId = self::$idCounter;
        $this->isOrganic = $isOrganic;

        if ($formula !== null) {
            $this->formula = $formula;
        } elseif (!empty($this->atoms)) {
            $this->computeFormula();
        } else {
            $this->formula = '';
        }
    }

    public static function resetIdCounter(): void
    {
        self::$idCounter = 0;
    }

    private function computeFormula(): void
    {
        $counts = [];
        foreach ($this->atoms as $atom) {
            $sym = $atom->symbol();
            $counts[$sym] = ($counts[$sym] ?? 0) + 1;
        }

        // Standard chemistry ordering: C, H, then alphabetical
        $parts = [];
        foreach (['C', 'H'] as $sym) {
            if (isset($counts[$sym])) {
                $n = $counts[$sym];
                $parts[] = $n > 1 ? "{$sym}{$n}" : $sym;
                unset($counts[$sym]);
            }
        }
        ksort($counts);
        foreach ($counts as $sym => $n) {
            $parts[] = $n > 1 ? "{$sym}{$n}" : $sym;
        }
        $this->formula = implode('', $parts);

        // Check if organic
        $hasC = false;
        $hasH = false;
        foreach ($this->atoms as $a) {
            if ($a->atomicNumber === 6) $hasC = true;
            if ($a->atomicNumber === 1) $hasH = true;
        }
        $this->isOrganic = $hasC && $hasH;
    }

    public function molecularWeight(): float
    {
        $total = 0.0;
        foreach ($this->atoms as $a) {
            $total += $a->massNumber;
        }
        return $total;
    }

    public function atomCount(): int
    {
        return count($this->atoms);
    }
}

class ChemicalReaction
{
    /**
     * @param string[] $reactants
     * @param string[] $products
     */
    public function __construct(
        public readonly array $reactants,
        public readonly array $products,
        public readonly float $activationEnergy = 1.0,
        public readonly float $deltaH = 0.0,
        public readonly string $name = '',
    ) {}

    /** Check if reaction can proceed at given temperature. */
    public function canProceed(float $temperature): bool
    {
        $thermalEnergy = K_B * $temperature;
        if ($thermalEnergy <= 0) {
            return false;
        }
        $rate = exp(-$this->activationEnergy / $thermalEnergy);
        return self::rand() < $rate;
    }

    private static function rand(): float
    {
        return mt_rand() / (mt_getrandmax() + 1);
    }
}

class ChemicalSystem
{
    /** @var Molecule[] */
    public array $molecules = [];

    public int $reactionsOccurred = 0;
    public int $waterCount = 0;
    public int $aminoAcidCount = 0;
    public int $nucleotideCount = 0;

    public function __construct(
        public readonly AtomicSystem $atomic,
    ) {}

    /** Form water molecules: 2H + O -> H2O */
    public function formWater(): array
    {
        $waters = [];
        $hydrogens = [];
        $oxygens = [];

        foreach ($this->atomic->atoms as $a) {
            if ($a->atomicNumber === 1 && empty($a->bonds)) {
                $hydrogens[] = $a;
            } elseif ($a->atomicNumber === 8 && count($a->bonds) < 2) {
                $oxygens[] = $a;
            }
        }

        while (count($hydrogens) >= 2 && !empty($oxygens)) {
            $h1 = array_pop($hydrogens);
            $h2 = array_pop($hydrogens);
            $o = array_pop($oxygens);

            $water = new Molecule(
                atoms: [$h1, $h2, $o],
                name: 'water',
            );
            $waters[] = $water;
            $this->molecules[] = $water;
            $this->waterCount++;

            // Form bonds
            $h1->bonds[] = $o->atomId;
            $h2->bonds[] = $o->atomId;
            $o->bonds[] = $h1->atomId;
            $o->bonds[] = $h2->atomId;
        }

        return $waters;
    }

    /** Form methane: C + 4H -> CH4 */
    public function formMethane(): array
    {
        $methanes = [];
        $carbons = [];
        $hydrogens = [];

        foreach ($this->atomic->atoms as $a) {
            if ($a->atomicNumber === 6 && empty($a->bonds)) {
                $carbons[] = $a;
            } elseif ($a->atomicNumber === 1 && empty($a->bonds)) {
                $hydrogens[] = $a;
            }
        }

        while (!empty($carbons) && count($hydrogens) >= 4) {
            $c = array_pop($carbons);
            $hs = [];
            for ($i = 0; $i < 4; $i++) {
                $hs[] = array_pop($hydrogens);
            }

            $methane = new Molecule(
                atoms: array_merge([$c], $hs),
                name: 'methane',
            );
            $methanes[] = $methane;
            $this->molecules[] = $methane;

            foreach ($hs as $h) {
                $c->bonds[] = $h->atomId;
                $h->bonds[] = $c->atomId;
            }
        }

        return $methanes;
    }

    /** Form ammonia: N + 3H -> NH3 */
    public function formAmmonia(): array
    {
        $ammonias = [];
        $nitrogens = [];
        $hydrogens = [];

        foreach ($this->atomic->atoms as $a) {
            if ($a->atomicNumber === 7 && empty($a->bonds)) {
                $nitrogens[] = $a;
            } elseif ($a->atomicNumber === 1 && empty($a->bonds)) {
                $hydrogens[] = $a;
            }
        }

        while (!empty($nitrogens) && count($hydrogens) >= 3) {
            $n = array_pop($nitrogens);
            $hs = [];
            for ($i = 0; $i < 3; $i++) {
                $hs[] = array_pop($hydrogens);
            }

            $ammonia = new Molecule(
                atoms: array_merge([$n], $hs),
                name: 'ammonia',
            );
            $ammonias[] = $ammonia;
            $this->molecules[] = $ammonia;

            foreach ($hs as $h) {
                $n->bonds[] = $h->atomId;
                $h->bonds[] = $n->atomId;
            }
        }

        return $ammonias;
    }

    /** Form an amino acid from available atoms. */
    public function formAminoAcid(string $aaType = 'Gly'): ?Molecule
    {
        $carbons = [];
        $hydrogens = [];
        $oxygens = [];
        $nitrogens = [];

        foreach ($this->atomic->atoms as $a) {
            if ($a->atomicNumber === 6 && empty($a->bonds)) {
                $carbons[] = $a;
            } elseif ($a->atomicNumber === 1 && empty($a->bonds)) {
                $hydrogens[] = $a;
            } elseif ($a->atomicNumber === 8 && count($a->bonds) < 2) {
                $oxygens[] = $a;
            } elseif ($a->atomicNumber === 7 && empty($a->bonds)) {
                $nitrogens[] = $a;
            }
        }

        if (count($carbons) < 2 || count($hydrogens) < 5
            || count($oxygens) < 2 || count($nitrogens) < 1) {
            return null;
        }

        $atoms = [];
        for ($i = 0; $i < 2; $i++) $atoms[] = array_pop($carbons);
        for ($i = 0; $i < 5; $i++) $atoms[] = array_pop($hydrogens);
        for ($i = 0; $i < 2; $i++) $atoms[] = array_pop($oxygens);
        $atoms[] = array_pop($nitrogens);

        $aa = new Molecule(
            atoms: $atoms,
            name: $aaType,
            isOrganic: true,
            functionalGroups: ['amino', 'carboxyl'],
        );
        $this->molecules[] = $aa;
        $this->aminoAcidCount++;

        // Form internal bonds
        $first = $atoms[0];
        for ($i = 1; $i < count($atoms); $i++) {
            $first->bonds[] = $atoms[$i]->atomId;
            $atoms[$i]->bonds[] = $first->atomId;
        }

        return $aa;
    }

    /** Form a nucleotide (sugar + phosphate + base). */
    public function formNucleotide(string $base = 'A'): ?Molecule
    {
        $carbons = [];
        $hydrogens = [];
        $oxygens = [];
        $nitrogens = [];

        foreach ($this->atomic->atoms as $a) {
            if ($a->atomicNumber === 6 && empty($a->bonds)) {
                $carbons[] = $a;
            } elseif ($a->atomicNumber === 1 && empty($a->bonds)) {
                $hydrogens[] = $a;
            } elseif ($a->atomicNumber === 8 && count($a->bonds) < 2) {
                $oxygens[] = $a;
            } elseif ($a->atomicNumber === 7 && empty($a->bonds)) {
                $nitrogens[] = $a;
            }
        }

        // Simplified requirements
        if (count($carbons) < 5 || count($hydrogens) < 8
            || count($oxygens) < 4 || count($nitrogens) < 2) {
            return null;
        }

        $atoms = [];
        for ($i = 0; $i < 5; $i++) $atoms[] = array_pop($carbons);
        for ($i = 0; $i < 8; $i++) $atoms[] = array_pop($hydrogens);
        for ($i = 0; $i < 4; $i++) $atoms[] = array_pop($oxygens);
        for ($i = 0; $i < 2; $i++) $atoms[] = array_pop($nitrogens);

        $nuc = new Molecule(
            atoms: $atoms,
            name: "nucleotide-{$base}",
            isOrganic: true,
            functionalGroups: ['sugar', 'phosphate', 'base'],
        );
        $this->molecules[] = $nuc;
        $this->nucleotideCount++;

        $first = $atoms[0];
        for ($i = 1; $i < count($atoms); $i++) {
            $first->bonds[] = $atoms[$i]->atomId;
            $atoms[$i]->bonds[] = $first->atomId;
        }

        return $nuc;
    }

    /** Run catalyzed reactions to form complex molecules. */
    public function catalyzedReaction(float $temperature, bool $catalystPresent = false): int
    {
        $formed = 0;
        $eaFactor = $catalystPresent ? 0.3 : 1.0;
        $thermal = K_B * $temperature;

        // Try to form amino acids
        if ($thermal > 0 && count($this->atomic->atoms) > 10) {
            $aaProb = exp(-5.0 * $eaFactor / ($thermal + 1e-20));
            if (self::rand() < $aaProb) {
                $aaType = AMINO_ACIDS[array_rand(AMINO_ACIDS)];
                if ($this->formAminoAcid($aaType) !== null) {
                    $formed++;
                    $this->reactionsOccurred++;
                }
            }
        }

        // Try to form nucleotides
        if ($thermal > 0 && count($this->atomic->atoms) > 19) {
            $nucProb = exp(-8.0 * $eaFactor / ($thermal + 1e-20));
            if (self::rand() < $nucProb) {
                $bases = ['A', 'T', 'G', 'C'];
                $base = $bases[array_rand($bases)];
                if ($this->formNucleotide($base) !== null) {
                    $formed++;
                    $this->reactionsOccurred++;
                }
            }
        }

        return $formed;
    }

    /** Count molecules by type. */
    public function moleculeCensus(): array
    {
        $counts = [];
        foreach ($this->molecules as $m) {
            $key = $m->name ?: $m->formula;
            $counts[$key] = ($counts[$key] ?? 0) + 1;
        }
        return $counts;
    }

    public function toSnapshot(): array
    {
        return [
            'molecule_count'     => count($this->molecules),
            'molecules_by_type'  => $this->moleculeCensus(),
            'water_count'        => $this->waterCount,
            'amino_acid_count'   => $this->aminoAcidCount,
            'nucleotide_count'   => $this->nucleotideCount,
            'reactions_occurred' => $this->reactionsOccurred,
        ];
    }

    private static function rand(): float
    {
        return mt_rand() / (mt_getrandmax() + 1);
    }
}
